### PR TITLE
Remove sender from invoke method

### PR DESCRIPTION
**Pull request by Tommo-L**
Fix https://github.com/neo-project/neo-modules/issues/351

### CODE REVIEW

**Review by shargon — APPROVED** [review]

**Review by devhawk — CHANGES_REQUESTED** [review]
Can we get our story straight on the sender parameter? @cloud8little just asked me to add the parameter to the new async RpcCLient PR: https://github.com/neo-project/neo-modules/pull/335#discussion_r502257495

**Comment by Tommo-L** [comment]
> Can we get our story straight on the sender parameter? @cloud8little just asked me to add the parameter to the new async RpcCLient PR: [#335 (comment)](https://github.com/neo-project/neo-modules/pull/335#discussion_r502257495)

`Signers[0]` is the sender.

**Comment by cloud8little** [comment]
why we make this change https://github.com/neo-project/neo-modules/pull/309 before?

**Comment by cloud8little** [comment]
> Can we get our story straight on the sender parameter? @cloud8little just asked me to add the parameter to the new async RpcCLient PR: [#335 (comment)](https://github.com/neo-project/neo-modules/pull/335#discussion_r502257495)

@devhawk  I got it clear from https://github.com/neo-project/neo-modules/pull/368#pullrequestreview-505857571,  no need to add sender.

**Comment by superboyiii** [comment]
> > Can we get our story straight on the sender parameter? @cloud8little just asked me to add the parameter to the new async RpcCLient PR: [#335 (comment)](https://github.com/neo-project/neo-modules/pull/335#discussion_r502257495)
> 
> `Signers[0]` is the sender.

Why `Signers[0]` should be the sender? `Signers[1]`, `Signers[2]` could also be sender, that's why we have to confirm the sender in params.

**Comment by superboyiii** [comment]
Due to my test for this PR:
1. if sender is `Signers[0]` and `Signers[0]` doesn't have any `GAS`, then it will choose `Signers[1]` as sender(if `Signers[1]` has enough GAS) when `MakeTransaction`. 
2. if `Signers[0]` and `Signers[1]` both have enough GAS, sender will be `Signers[0]` even if I set `Signers[1]` as `Signer`.
***I think this change would make `sender` more confusion because we want `Signers[0]` to be sender, but in fact it still can't always be like this. Then why we have to set `Signers[0]` to be sender? Developer has their authority to choose which is the sender.***

**Comment by roman-khimov** [comment]
> Developers have their authority to choose which is the sender.

Absolutely if you're to transfer signers given to "invoke*" into transaction 1:1. They'd get exactly the sender/signers they want this way.

**Comment by superboyiii** [comment]
> > Developers have their authority to choose which is the sender.
> 
> Absolutely if you're to transfer signers given to "invoke*" into transaction 1:1. They'd get exactly the sender/signers they want this way.

Now I agree with you, just wait for solving https://github.com/neo-project/neo-modules/issues/370

**Comment by superboyiii** [comment]
Let's make a summary:
1. `sender` should be from the input signer[].
2. If the input `signer[0]` hasn't enough GAS for make Tx, it could use `signer[1]` to be sender if it exist in the input signer[] and it has enough GAS as well. Or an error msg should be returned such as "Insufficient GAS".
3. `Signer[]` and `sender` should be from the input params but not from `wallet` class, should never automatically choosing an address from `wallet` even if the input `Signer[]` doesn't have enough GAS, the response should always be from what we input.

**Comment by roman-khimov** [comment]
Just leave the first point and that's it. Second one is not necessary, it's up to the user of this API to order signers correctly, so if `signer[0]` doesn't have enough GAS --- it's not a problem of RPC server. Think also of RPC servers that don't manage wallets like C# one does, they just do 1:1 mapping from `invoke*` signers into transaction signers and don't care about balances at all, it's not their problem.

**Comment by superboyiii** [comment]
> Just leave the first point and that's it. Second one is not necessary, it's up to the user of this API to order signers correctly, so if `signer[0]` doesn't have enough GAS --- it's not a problem of RPC server. Think also of RPC servers that don't manage wallets like C# one does, they just do 1:1 mapping from `invoke*` signers into transaction signers and don't care about balances at all, it's not their problem.

But current neo-cli is designed to care, we should solve this conflict in `wallet.cs`. Current `MakeTransaction` will check balance when invoking invokefunction, I think a new `MakeTransaction` should be created for this.

**Comment by superboyiii** [comment]
Merged with latest, find it work well when signers[0] doesn't have enough gas.
```
{
  "jsonrpc": "2.0",
  "method": "invokefunction",
  "params": [
    "0x254b9decd76080ef368e7a6b0a065938dfbc31cf",
    "balanceOf",
    [
      {
        "type": "Hash160",
        "value": "0xf642a0401fdd56a03114639a98b7963eb587a30a"
      }
    ],
    [
      {
        "account": "0x7170793e96092211675c8333395d79c465a4f56f",
        "scopes": "CalledByEntry"
      },
      {
        "account": "0xf642a0401fdd56a03114639a98b7963eb587a30a",
        "scopes": "CalledByEntry"
      }
    ]
  ],
  "id": 3
}
```

```
{
    "jsonrpc": "2.0",
    "id": 3,
    "result": {
        "script": "DBQKo4e1Ppa3mJpjFDGgVt0fQKBC9hHADAliYWxhbmNlT2YMFM8xvN84WQYKa3qONu+AYNfsnUslQWJ9W1I=",
        "state": "HALT",
        "gasconsumed": "3840690",
        "exception": "Insufficient GAS",
        "stack": [
            {
                "type": "Integer",
                "value": "2000000000000000"
            }
        ]
    }
}
```

**Review by superboyiii — APPROVED** [review]

**Review by devhawk — APPROVED** [review]
👍

File: src/RpcServer/RpcServer.SmartContract.cs [hunk 1]
```diff
@@ -103,23 +103,22 @@ protected virtual JObject InvokeFunction(JArray _params)
...
         }
 
         [RpcMethod]
         protected virtual JObject InvokeScript(JArray _params)
         {
-            byte[] script = Convert.FromBase64String(_params[0].AsString());
-            UInt160 sender = _params.Count >= 2 ? AddressToScriptHash(_params[1].AsString()) : null;
-            Signers signers = _params.Count >= 3 ? SignersFromJson((JArray)_params[2]) : null;
-            return GetInvokeResult(script, sender, signers);
+            byte[] script = _params[0].AsString().HexToBytes();
```
Comment (ProDog) [inline src/RpcServer/RpcServer.SmartContract.cs:119]: The modification of https://github.com/neo-project/neo-modules/pull/362 goes back here 🤔

**Pull request merged**